contracts/staking/GoodFundManager.sol

```diff
@@ -189,76 +237,127 @@ contract GoodFundManager is DAOContract {
+        uint256 gdAmountToMint= getGasPriceInGD(totalUsedGas);
+        GoodReserveCDai(nameService.addresses(nameService.RESERVE())).mintRewardFromRR(nameService.getAddress("CDAI"),msg.sender,gdAmountToMint);
+        lastCollectedInterest = block.timestamp;
         emit FundsTransferred(
             msg.sender,
-            address(_staking),
-            reserveAddress,
+            nameService.addresses(nameService.RESERVE()),
             interest,
-            gdUBI
+            gdUBI,
+            gdAmountToMint
         );
     }
+
+    /**
+     * @dev  Function that get addresses of staking contracts which interests 
+     * can be collected with the gas amount that provided as parameter
+     * @param _maxGasAmount The maximum amount of the gas that keeper willing to spend collect interests
+     * @return addresses of the staking contracts to the collect interests
+     */
+    function calcSortedContracts(uint256 _maxGasAmount) public view returns(address[] memory){
+        uint activeContractsLength = activeContracts.length;
+        address[] memory addresses = new address[](activeContractsLength); 
+        uint256[] memory balances = new uint256[](activeContractsLength);
+        uint256 tempInterest;
+        uint totalInterest;
+        uint i;
+        require(activeContractsLength > 0 , "There should be at least one active staking contract");
+        for (i = 0; i < activeContractsLength; i++){
+            (tempInterest, ,) = StakingContract(activeContracts[i]).currentUBIInterest();
+            totalInterest += tempInterest;
+            if (tempInterest != 0){
+                addresses[i] = activeContracts[i];
+                balances[i] = tempInterest;
+            }
+        }
+        uint gasCostInCdai = getGasPriceInCDAI(_maxGasAmount); // Get gas price in cDAI so can compare with possible interest amount to get
+        require(totalInterest >= gasCostInCdai,"Current interest's worth less than spent gas worth");
+        quick(balances,addresses); // sort the values according to interest balance
+        uint gasCost;
+        uint possibleCollected;
+        for(i = activeContractsLength - 1; i >= 0; i--){ // elements are sorted by balances from lowest to highest 
+		
+            if(addresses[i] != address(0x0)){
+                gasCost = StakingContract(addresses[i]).getGasCostForInterestTransfer();
+                if(_maxGasAmount - gasCost >= 650000){ // this value will change. Its hardcoded for further transactions such as ubiMINTING,gas price calculations and gdMINT
+                    // collects the interest from the staking contract and transfer it directly to the reserve contract
+                    //`collectUBIInterest` returns (iTokengains, tokengains, precission loss, donation ratio)
+					possibleCollected += balances[i];
+                    _maxGasAmount = _maxGasAmount - gasCost;
+                }else{
+					addresses[i] = address(0x0);
```

Comment:
i think we can break here also.
then nullify all address slots <i (saves the call to getCastCost for each following contract)

Reply:
Changed